overlord/ifacestate/handlers.go

```diff
@@ -22,33 +22,83 @@ package ifacestate
 import (
 	"errors"
 	"fmt"
+	"path"
 	"reflect"
 	"sort"
 	"strings"
 	"time"
 
 	"gopkg.in/tomb.v2"
 
+	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/i18n"
 	"github.com/snapcore/snapd/interfaces"
 	"github.com/snapcore/snapd/interfaces/hotplug"
 	"github.com/snapcore/snapd/logger"
 	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/servicestate"
 	"github.com/snapcore/snapd/overlord/snapstate"
 	"github.com/snapcore/snapd/overlord/state"
 	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/quota"
 	"github.com/snapcore/snapd/timings"
 )
 
 var snapstateFinishRestart = snapstate.FinishRestart
 
+// addJournalQuotaLayout handles the addition of a journal quota bind mount
+// to mimicks what systemd does for services with log namespaces.
+func addJournalQuotaLayout(quotaGroup *quota.Group, layouts *[]snap.Layout) error {
+	if quotaGroup.JournalLimit == nil {
+		return nil
+	}
+
+	// bind mount the journal namespace folder ontop of the journal folder
+	// /etc/systemd/journal.<ns> -> /etc/systemd/journal
+	journalLayout := snap.Layout{
+		Bind: path.Join(dirs.SnapSystemdDir, fmt.Sprintf("journal.snap-%s", quotaGroup.Name)),
+		Path: path.Join(dirs.SnapSystemdDir, "journal"),
+		Mode: 0755,
+	}
+	*layouts = append(*layouts, journalLayout)
+	return nil
+}
+
+// getExtraLayouts helper function to dynamically calculate the extra mount layouts for
+// a snap instance. These are the layouts which can change during the lifetime of a snap
+// like for instance mimicking systemd journal namespace mount layouts.
+func getExtraLayouts(st *state.State, snapInstanceName string) ([]snap.Layout, error) {
+	snapOpts, err := servicestate.SnapServiceOptions(st, snapInstanceName, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var extraLayouts []snap.Layout
+	if snapOpts.QuotaGroup != nil {
+		if err := addJournalQuotaLayout(snapOpts.QuotaGroup, &extraLayouts); err != nil {
+			return extraLayouts, err
+		}
+	}
+
+	return extraLayouts, nil
+}
+
 // confinementOptions returns interfaces.ConfinementOptions from snapstate.Flags.
-func confinementOptions(flags snapstate.Flags) interfaces.ConfinementOptions {
+func confinementOptions(flags snapstate.Flags, extraLayouts []snap.Layout) interfaces.ConfinementOptions {
 	return interfaces.ConfinementOptions{
-		DevMode:  flags.DevMode,
-		JailMode: flags.JailMode,
-		Classic:  flags.Classic,
+		DevMode:      flags.DevMode,
+		JailMode:     flags.JailMode,
+		Classic:      flags.Classic,
+		ExtraLayouts: extraLayouts,
+	}
+}
+
+func buildConfinementOptions(st *state.State, snapInstanceName string, flags snapstate.Flags) interfaces.ConfinementOptions {
+	extraLayouts, err := getExtraLayouts(st, snapInstanceName)
+	if err != nil {
+		logger.Noticef("cannot get extra mount layouts of snap %q: %s", snapInstanceName, err)
```

Comment:
I intentionally 'swallow' the exception here because I felt like robustness would be better than correctness for these dynamically calculated mount additions. I'm writing this comment just to explain that this was intentional, and that in my head it makes a lot more sense to ignore should anything go wrong here and not stop the entire process, and then we log this instead. I will change this if any of you feel strongly otherwise about this.

Reply:
we almost never reason that way, we just propagate errors if things can start to fail. There is nothing done by getExtraLayouts that depends on fragile things we don't control? or am I missing something

Reply:
Thanks for your input here @pedronis, I'll change it to to propegate the error instead